OpenRA.Mods.Common/Activities/Rearm.cs

```diff
@@ -8,53 +8,64 @@
  */
 #endregion
 
+using System.Collections.Generic;
 using System.Linq;
 using OpenRA.Activities;
 using OpenRA.Mods.Common.Traits;
-using OpenRA.Traits;
 
 namespace OpenRA.Mods.Common.Activities
 {
 	public class Rearm : Activity
 	{
-		readonly LimitedAmmo limitedAmmo;
-		int ticksPerPip = 25 * 2;
-		int remainingTicks = 25 * 2;
-		string sound = null;
+		readonly IEnumerable<AmmoPool> ammoPools;
+		readonly Dictionary<AmmoPool, int> ammoPoolsReloadTimes;
 
-		public Rearm(Actor self, string sound = null)
+		public Rearm(Actor self)
 		{
-			limitedAmmo = self.TraitOrDefault<LimitedAmmo>();
-			if (limitedAmmo != null)
-				ticksPerPip = limitedAmmo.ReloadTimePerAmmo();
-			remainingTicks = ticksPerPip;
-			this.sound = sound;
+			ammoPools = self.TraitsImplementing<AmmoPool>();
+
+			if (ammoPools == null)
+				return;
+
+			ammoPoolsReloadTimes = ammoPools.ToDictionary(x => x, y => y.Info.ReloadTicks);
 		}
 
 		public override Activity Tick(Actor self)
 		{
-			if (IsCanceled || limitedAmmo == null)
+			if (IsCanceled || ammoPoolsReloadTimes == null)
 				return NextActivity;
 
-			if (--remainingTicks == 0)
+			var needsReloading = false;
+
+			foreach (var pool in ammoPools)
 			{
-				var hostBuilding = self.World.ActorMap.GetUnitsAt(self.Location)
-					.FirstOrDefault(a => a.HasTrait<RenderBuilding>());
+				if (pool.FullAmmo())
+					continue;
+
+				needsReloading = true;
+
+			    if (--ammoPoolsReloadTimes[pool] > 0)
+			        continue;
+
+				// HACK to check if we are on the helipad/airfield/etc.
+				var hostBuilding = self.World.ActorMap.GetUnitsAt(self.Location).FirstOrDefault(a => a.HasTrait<RenderBuilding>());
 
 				if (hostBuilding == null || !hostBuilding.IsInWorld)
 					return NextActivity;
 
-				if (!limitedAmmo.GiveAmmo())
-					return NextActivity;
+				if (!pool.GiveAmmo())
+					continue;
 
 				hostBuilding.Trait<RenderBuilding>().PlayCustomAnim(hostBuilding, "active");
```

Comment:
It's actually better to play the anim for each ammo pool anyway.
When we refactor rearming properly in a follow-up (which includes getting rid of this ugly building hack), we can actually use this to play different animations for each ammo pool (by replacing the "active" hardcoding with something customisable), so players can get a visual indicator what exactly has been reloaded just now.